### PR TITLE
Track message status by channel especially when encountering exceptions

### PR DESCRIPTION
**Pull request by vrajmohan**

### CODE REVIEW

**Review by mpidcock — APPROVED** [review]
Nice. I like that `UnsubscribedException` extends `MessageSendException`. Out of curiosity, I sent an email without correct credentials, and the `MessageSendException` exception was thrown as expected with `[401 Unauthorized] during [POST] to ...` and recorded on the message 🙌

File: src/main/java/org/codeforamerica/messaging/repositories/MessageRepository.java [hunk 1]
```diff
@@ -13,18 +13,16 @@ public interface MessageRepository extends CrudRepository<Message, Long> {
...
-             SUM(CASE sms.status WHEN 'delivered' THEN 1 ELSE 0 END) AS deliveredSmsCount,
-             SUM(CASE sms.status WHEN 'undelivered' THEN 1 ELSE 0 END) AS undeliveredSmsCount
+             SUM(CASE emailStatus WHEN 'accepted' THEN 1 ELSE 0 END) AS acceptedEmailCount,
+             SUM(CASE emailStatus WHEN 'rejected' THEN 1 ELSE 0 END) AS rejectedEmailCount,
+             SUM(CASE emailStatus WHEN 'delivered' THEN 1 ELSE 0 END) AS deliveredEmailCount,
+             SUM(CASE emailStatus WHEN 'failed' THEN 1 ELSE 0 END) AS undeliveredEmailCount,
+             SUM(CASE smsStatus WHEN 'accepted' THEN 1 ELSE 0 END) AS acceptedSmsCount,
+             SUM(CASE smsStatus WHEN 'rejected' THEN 1 ELSE 0 END) AS rejectedSmsCount,
+             SUM(CASE smsStatus WHEN 'delivered' THEN 1 ELSE 0 END) AS deliveredSmsCount,
+             SUM(CASE smsStatus WHEN 'undelivered' THEN 1 ELSE 0 END) AS undeliveredSmsCount
```
Comment (mpidcock) [inline src/main/java/org/codeforamerica/messaging/repositories/MessageRepository.java:23]: Just spotted this inconsistency. Slight preference for `failed` in both over `undelivered`.
> 'failed' AS undeliveredEmailCount
> 'undelivered' AS undeliveredSmsCount

File: src/main/java/org/codeforamerica/messaging/services/MessageService.java [hunk 2]
```diff
@@ -111,9 +112,13 @@ public void sendMessage(Long messageId) {
                 String smsBody = templateVariant.build(TemplateVariant::getSmsBody, templateParams);
                 SmsMessage sentSmsMessage = this.smsService.sendSmsMessage(message.getToPhone(), smsBody);
                 message.setSmsMessage(sentSmsMessage);
+                message.setSmsStatus("provider_accepted");
```
[the remark this answers is not in the document]
Comment (vrajmohan) [reply src/main/java/org/codeforamerica/messaging/services/MessageService.java:115]: I've been wondering about that as well. Perhaps the time is ripe. I would suggest that we don't hold up this PR but add a chore for doing it ASAP.

File: src/test/java/org/codeforamerica/messaging/controllers/MessageControllerTest.java [hunk 3]
```diff
@@ -43,32 +44,12 @@ public void getMessageUnauthenticated() throws Exception {
...
-                        .toPhone(TestData.TO_PHONE).build()));
-
-        mockMvc.perform(get("/api/v1/messages/" + TestData.BASE_ID))
-                .andExpect(MockMvcResultMatchers.status().isOk())
-                .andExpect((MockMvcResultMatchers.content().json(expectedResponse)));
-    }
-
-    @Test
-    @WithMockUser
+    @Disabled("TODO: This should be set to api_success or similar")
```
[the remark this answers is not in the document]
Comment (vrajmohan) [reply src/test/java/org/codeforamerica/messaging/controllers/MessageControllerTest.java:47]: Thanks for catching it. This was a TODO when the test originally failed. The test itself is useless and is replaced by `MessageServiceErrorHandlingTest`.

File: src/main/java/org/codeforamerica/messaging/repositories/MessageRepository.java (hunk 1 above)
Comment (vrajmohan) [reply src/main/java/org/codeforamerica/messaging/repositories/MessageRepository.java:23]: Great catch. The issue is that `failed` is returned by `Mailgun` for undelivered emails and this query deals with that anomaly. However, I introduced `failed` in this story for provider handoff failures without realizing that it was already used. I should replace the new `failed` with `provider_rejected` or similar :-(

**Review by mpidcock — APPROVED** [review]

**Pull request merged**